### PR TITLE
Verbose exception messages for missing/unknown parameters in QueryBuilder

### PR DESCRIPTION
**Pull request by svencan**

### CODE REVIEW

**Comment by svencan** [comment]
Based on this request: #5609

File: lib/Doctrine/ORM/Query.php [hunk 1]
```diff
@@ -306,12 +306,8 @@ protected function doExecute()
         $paramCount    = count($this->parameters);
         $mappingCount  = count($paramMappings);
 
-        if ($paramCount > $mappingCount) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Query.php]: This should remain?

Comment (svencan) [reply lib/Doctrine/ORM/Query.php]: Setting more parameters than are mapped raises `QueryException::unknownParameter`:

https://github.com/doctrine/doctrine2/blob/695edb9fdcafbe81f3923317b1e9c5b4355be308/lib/Doctrine/ORM/Query.php#L396-L398

File: lib/Doctrine/ORM/Query/QueryException.php [hunk 2]
```diff
@@ -116,6 +94,24 @@ public static function unknownParameter($key)
...
+    public static function missingParameter($expected, $received)
+    {
+        foreach (array_keys($expected) as $token) {
+            if (!in_array($token, $received)) {
+                return new self('Missing parameter: no parameter set for token ' . $token . '.');
+            }
+        }
+        
+        return new self('Too few parameters: the query defines ' . count($expected)
+                . ' parameters but you only bound ' . count($received));
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Query/QueryException.php]: Would it be feasible to make an exception that uses a `json_serialize` of expected vs passed parameters (plus the diff)?

Comment (Ocramius) [reply lib/Doctrine/ORM/Query/QueryException.php]: To clarify, from a dev perspective, if I forget 2 params and do a "edit and refresh until it works", this kind of exception on a single parameter means multiple refreshes on multiple mistakes

File: lib/Doctrine/ORM/Query/QueryException.php [hunk 3]
```diff
@@ -75,28 +75,6 @@ public static function invalidParameterPosition($pos)
         return new self('Invalid parameter position: ' . $pos);
     }
 
-    /**
-     * @param int $expected
-     * @param int $received
-     *
-     * @return QueryException
-     */
-    public static function tooManyParameters($expected, $received)
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Query/QueryException.php]: @greg0ire exceptions are your thing: how much does this overlap with your patch?

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: lib/Doctrine/ORM/Query/QueryException.php (hunk 3 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Query/QueryException.php]: It doesn't: I restrained myself to ORMException for now.

Comment (Ocramius) [reply lib/Doctrine/ORM/Query/QueryException.php]: Thanks!

@svencan besides improvements to be applied as per https://github.com/doctrine/doctrine2/pull/7084#discussion_r170472606, this looks good :+1:

File: lib/Doctrine/ORM/Query/QueryException.php (hunk 2 above)
Comment (svencan) [reply lib/Doctrine/ORM/Query/QueryException.php]: @Ocramius can you elaborate on what you mean by `json_serialize`?

Comment (svencan) [reply lib/Doctrine/ORM/Query/QueryException.php]: I'm confused, do you mean `json_encode` or are you really referring to [JsonSerializable](http://php.net/manual/en/jsonserializable.jsonserialize.php)?

Comment (Ocramius) [reply lib/Doctrine/ORM/Query/QueryException.php]: Yeah, sorry, `json_encode()` - PHP's awesome naming.

**Comment by derrabus** [comment]
I'm closing this PR because it hasn't seen progress for several years. Please open a new one if you want to pick up the work one day.

**Pull request closed without merging**